### PR TITLE
add Optuna

### PR DESCRIPTION
**Pull request by savchenkoyana**
Hello, thank you for this amazing project!

We ([ENOT](https://enot.ai/) team) have implemented hyperparamenter search with [Optuna](https://optuna.org/) and want to share it with the community. Also there are other little changes in this PR:
- Add `data.yaml` dump to save_dir
- Add default albumentations augs to hyp configs so that users can modify them if they want without hardcode
- Add RAdam optimizer

Related to [issue comment](https://github.com/ultralytics/yolov5/issues/918#issuecomment-721240802)

Ciao,
Yana

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced image augmentations and introduced Optuna hyperparameter optimization in YOLOv5 training.

### 📊 Key Changes
- Added rotation with a certain probability to image augmentations.
- Integrated a variety of new albumentation augmentations such as blur, median blur, gray, CLAHE, brightness_contrast, random_gamma, and image compression.
- Created custom Optuna hyperparameter optimization with additional CLI arguments for pruner configuration and more.
- Introduced `trial_last` and `trial_best` subdirectories for saving checkpoint states during Optuna optimization.
- Accommodated TensorBoard logger to work with Optuna as well.
- Ensured other utilities like data loaders and augmentations now accept the new hyperparameters (`hyp`) configuration.

### 🎯 Purpose & Impact
- The purpose of these changes is to improve model performance by enhancing augmentation techniques and optimizing hyperparameters more efficiently.
- Impacts are expected in terms of improved model accuracy and generalizability due to more advanced augmentation strategies.
- Users now have the option for systematic hyperparameter tuning using Optuna, leading to potentially better performing models. 🎛️
- The addition of a rotation probability adds flexibility to model training, as not every image will be rotated, which can be beneficial depending on the dataset. 🔄
- The impact of albumentations is significant, as it adds modern, diverse image augmentation techniques that can enrich training datasets and improve model resilience. 🏞️
- The introduction of a structured approach for hyperparameter optimization with Optuna provides a robust alternative to manual or less sophisticated methods, which can be time-consuming and less effective. 🧬

### CODE REVIEW

**Comment by AyushExel** [comment]
@savchenkoyana thanks for the PR. I'm interested in exploring more hyper-parameter optimization options. We'll take a look at this after the next release. I have a few questions:
But I just wanted to confirm this whether this will break the existing evolve workflow?

**Comment by savchenkoyana** [comment]
> @savchenkoyana thanks for the PR. I'm interested in exploring more hyper-parameter optimization options. We'll take a look at this after the next release. I have a few questions: But I just wanted to confirm this whether this will break the existing evolve workflow?

We compared ultralytics genetics with Optuna and both procedures worked okay on this branch

**Comment by AyushExel** [comment]
@savchenkoyana perfect! Thanks

**Comment by elon1992** [comment]
Hello @savchenkoyana @AyushExel @glenn-jocher,

I want to do the hyper-parameter optimization using Optuna using Bayesian Optimization method. Could you please suggest me the changes to do and how to do it?
Thank you in advance.

**Comment by glenn-jocher** [comment]
@elon1992 hello! 👋 Exciting to see your interest in hyper-parameter optimization using Optuna with Bayesian Optimization! While the YOLOv5 repo currently doesn't contain a custom example for Optuna with Bayesian Optimization, integrating it should be fairly straightforward if you're familiar with Optuna's framework.

Here's a basic outline you can start with:
1. You'll need to define an objective function that trains your YOLOv5 model and returns a value to minimize or maximize, like validation loss. This function accepts a `trial` object from which you can suggest hyperparameters to test.
2. Use `optuna.create_study` method with the `direction` set based on whether you want to minimize or maximize your objective. For Bayesian Optimization, you can specify the `sampler` as `optuna.samplers.TPESampler()`, which is a Bayesian-based approach.
3. Invoke `study.optimize()`, passing your objective function and the number of trials.

Here's a tiny example to illustrate:

```python
import optuna

def objective(trial):
    # Define your hyperparameters using trial.suggest_*
    lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True)
    momentum = trial.suggest_float("momentum", 0.6, 0.99)
    # Add more parameters as needed

    # Your model training and evaluation logic here
    # For instance, train your model and return validation loss
    validation_loss = train_yolo(lr=lr, momentum=momentum)
    return validation_loss

study = optuna.create_study(direction="minimize", sampler=optuna.samplers.TPESampler())
study.optimize(objective, n_trials=100)

print("Best trial:", study.best_trial.params)
```

You'll need to adapt this to integrate with YOLOv5 training specifics, but it should provide a good starting point. Make sure to check Optuna's documentation for more detailed information on their Bayesian optimization capabilities and best practices.

Happy optimizing!
